include/picongpu/plugins/openPMD/openPMDWriter.hpp

```diff
@@ -591,6 +591,144 @@ Make sure that environment variable OPENPMD_BP_BACKEND is not set to ADIOS1.
+
+                params->initDataset<simDim>(
+                    mrc,
+                    ::openPMD::determineDatatype<ReinterpretedType>(),
+                    fieldsGlobalSizeDims,
+                    true,
+                    params->compressionMethod,
+                    datasetName);
+
+                // define record component level attributes
+                auto const inCellPosition = std::vector<float_X>(simDim, 0.0_X);
+                mrc.setPosition(inCellPosition);
+                auto const unit = 1.0_X;
+                mrc.setUnitSI(unit);
+
+                auto& buffer = rngProvider->getStateBuffer();
+                // getPointer() will wait for device->host transfer
+                ValueType* nativePtr = buffer.getHostBuffer().getPointer();
+                ReinterpretedType* rawPtr = reinterpret_cast<ReinterpretedType*>(nativePtr);
+                mrc.storeChunk(
+                    ::openPMD::shareRaw(rawPtr),
+                    asStandardVector(fieldsOffsetDims),
+                    asStandardVector(fieldsSizeDims));
+                params->openPMDSeries->flush();
+            }
+
+            /** Implementation of loading random number generator states
+             *
+             * Note: LoadFields cannot be easily used inside this function, since states are custom types.
+             * Instead, we load states as a byte (char) array and reinterpret.
+             * This matches how they are written in writeRngStates().
+             *
+             * Only suitable for the currently implemented RNG storage: one state per cell, no guards.
+             * If this doesn't hold, the implementation can't work as is, and an exception will be thrown.
+             */
+            HINLINE void loadRngStatesImpl(ThreadParams* params)
+            {
+                DataConnector& dc = Environment<simDim>::get().DataConnector();
+                using RNGProvider = pmacc::random::RNGProvider<simDim, random::Generator>;
+                auto rngProvider = dc.get<RNGProvider>(RNGProvider::getName());
+                auto const name = rngProvider->getName();
+
+                ::openPMD::Iteration iteration = params->openPMDSeries->WRITE_ITERATIONS[params->currentStep];
+                ::openPMD::Mesh mesh = iteration.meshes[name];
+                ::openPMD::MeshRecordComponent mrc = mesh[::openPMD::RecordComponent::SCALAR];
+
+                const pmacc::Selection<simDim> localDomain = Environment<simDim>::get().SubGrid().getLocalDomain();
+                auto fieldsSizeDims = params->window.localDimensions.size;
+                auto fieldsOffsetDims = localDomain.offset;
+
+                if(fieldsSizeDims != rngProvider->getSize())
+                    throw std::runtime_error("openPMD: RNG state can't be loaded due to not matching size");
+
+                // Reinterpret state as chars, it must be bitwise-copyable for it
+                using ReinterpretedType = char;
```

Comment:
Note that this kills portability, cross-platform as well as cross-application. The files written by the openPMD plugin are generally self-contained, but this field would be merely a raw byte-by-byte serialization. Is it an option to reinterpret_cast to an underlying machine type, e.g. by requiring that `RNGProvider::Buffer::ValueType` has an associated `UnderlyingType`?

Reply:
Cross-application is not needed here, it is only for checkpointing (and optional at that). Cross-platform and cross-application would be desired of course. The issue is I do not have a definition of this type, as it may come from a number of 3rd party libraries. The implementations I see have no fundamental underlying type, and are some structs with several members. So to me it appeared as the only solution.
I've actually [asked in the openPMD API repo about how to approach it](https://github.com/openPMD/openPMD-api/issues/1098).

Reply:
I was expecting this answer, I just wanted to mention it. In the given case, I think it would be the easiest solution for now.